torch/csrc/distributed/c10d/intra_node_comm.cu

```diff
@@ -0,0 +1,707 @@
+  }
+  X(2);
+  X(3);
+  X(4);
+  X(5);
+  X(6);
+  X(7);
+  X(8);
+#undef X
+
+  if (output.data_ptr() != input.data_ptr()) {
+    AT_CUDA_CHECK(cudaMemcpyAsync(
+        input.data_ptr(),
+        output.data_ptr(),
+        input.numel() * input.element_size(),
+        cudaMemcpyDeviceToDevice,
+        stream));
+  }
+  return input;
+}
+
+at::Tensor hybridCubeMeshAllReduce(
+    const at::Tensor& input,
+    std::array<void*, kMaxDevices> p2pStates,
+    std::array<void*, kMaxDevices> buffers,
+    int hcmInfo[4],
+    size_t rank,
+    size_t worldSize,
+    at::cuda::CUDAStream& stream) {
+  checkInput(input, rank);
+
+  size_t numelPerWarp = kBytesPerThread / input.element_size() * kWarpSize;
+  size_t N_aligned = alignUp(input.numel(), numelPerWarp);
+  TORCH_CHECK(N_aligned <= kMaxIntraNodeSize / input.element_size());
+
+  dim3 blocks, threads;
+  getLaunchConfig(N_aligned, input.element_size(), blocks, threads);
+
+  at::cuda::OptionalCUDAGuard guard(input.get_device());
+  AT_CUDA_CHECK(cudaMemcpyAsync(
+      buffers[rank],
+      input.data_ptr(),
+      input.numel() * input.element_size(),
+      cudaMemcpyDeviceToDevice,
+      stream));
+
+#define X(kAligned)                                                        \
+  hybridCubeMeshAllReduceKernel<kAligned><<<blocks, threads, 0, stream>>>( \
+      input.data_ptr<at::BFloat16>(),                                      \
+      input.numel(),                                                       \
+      N_aligned,                                                           \
+      castArr<P2pState*>(p2pStates),                                       \
+      castArr<at::BFloat16*>(buffers),                                     \
+      hcmInfo,                                                             \
+      rank);                                                               \
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+  if (N_aligned == static_cast<size_t>(input.numel())) {
+    X(true);
+  } else {
+    X(false);
+  }
+#undef X
+  return input;
+}
+
+AllReduceAlgo selectAllReduceAlgo(
+    const at::Tensor& input,
+    Topology topology,
+    size_t worldSize) {
+  // Only support bf16 for now
+  if (input.dtype() != at::kBFloat16) {
+    return AllReduceAlgo::NONE;
+  }
+  const auto numel = input.numel();
+  const auto numelPerWarp = kBytesPerThread / input.element_size() * kWarpSize;
+  if (topology == Topology::HYBRID_CUBE_MESH) {
+    TORCH_CHECK(
+        worldSize == 8, "hyperCubeAllReduce only supports exactly 8 GPUs");
+    if (alignUp(numel, numelPerWarp) <= kHcmThreshBytes) {
+      return AllReduceAlgo::HCM;
+    }
+  }
+  if (topology == Topology::FULLY_CONNECTED) {
+    if (alignUp(numel, numelPerWarp) <= kOneShotThreshBytes) {
+      return AllReduceAlgo::ONE_SHOT;
+    }
+    if (alignUp(numel, numelPerWarp * worldSize) <= kTwoShotThreshBytes) {
+      return AllReduceAlgo::TWO_SHOT;
+    }
+  }
+  return AllReduceAlgo::NONE;
+}
```

Comment:
Should we add unit tests to ensure that the correctness of the implementations don't regress, in case other people modify the code in this file or the cpp file in the future?